pip/req/req_set.py

```diff
@@ -286,6 +286,14 @@ def add_requirement(self, install_req, parent_req_name=None):
                                  existing_req, existing_req.extras)
                     # And now we need to scan this.
                     result = [existing_req]
+                elif not install_req.constraint:
```

Comment:
It'd be nice if you could rebase this; there's a few questions I wanna ask but I'll wait for a response from you before asking them.